src/snowflake/snowpark/dataframe_reader.py

```diff
@@ -1024,3 +1073,480 @@ def _read_semi_structured_file(self, path: str, format: str) -> DataFrame:
+                            )
+                            thread_future.add_done_callback(
+                                functools.partial(
+                                    ingestion_thread_cleanup_callback, file
+                                )
+                            )
+                            thread_pool_futures.append(thread_future)
+                            logger.debug(
+                                f"Submitted file {file} to thread executor for ingestion."
+                            )
+                        except queue.Empty:
+                            all_job_done = True
+                            unfinished_process_pool_futures = []
+                            logger.debug(
+                                "Parquet queue is empty, checking unfinished process pool futures."
+                            )
+                            for future in process_pool_futures:
+                                if future.done():
+                                    try:
+                                        future.result()  # Throw error if the process failed
+                                        logger.debug(
+                                            "A process future completed successfully."
+                                        )
+                                    except Exception as e:
+                                        logger.error(
+                                            f"Process future failed with error: {e}"
+                                        )
+                                        raise
+                                else:
+                                    unfinished_process_pool_futures.append(future)
+                                    all_job_done = False
+                            if all_job_done and parquet_file_queue.empty():
+                                # all jod is done and parquet file queue is empty, we finished all the fetch work
+                                # now we just need to wait for all ingestion threads to complete
+                                logger.debug(
+                                    "All jobs are done, and the parquet file queue is empty. Fetching work is complete."
+                                )
+                                break
+                            process_pool_futures = unfinished_process_pool_futures
+                            time.sleep(0.5)
+
+                    for future in as_completed(thread_pool_futures):
+                        try:
+                            future.result()  # Throw error if the thread failed
+                            logger.debug("A thread future completed successfully.")
+                        except BaseException as e:
+                            logger.error(f"Thread future failed with error: {e}")
+                            raise
+
+            except BaseException:
+                # graceful shutdown
+                process_executor.shutdown(wait=True)
```

Comment:
Is this for KeyboardInterrupted?

Reply:
this can cover KeyboardInterrupted, but mainly for errors happened during download/upload,
in current version, fail of one partition will fail the whole data, this code is meant to release resource under that situation